Autoloader: use all of the map files to populate the global maps

kbrown9 opened the pull request:
#### Changes proposed in this Pull Request:
* The latest autoloader needs to examine all of the classmap and filemap files created by the new autoloaders and populate the globals. In the `enqueue_files()` function, loop through all of the active plugins and examine the classmap and filemap files.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This changes an existing part of Jetpack.

#### Testing instructions:
* Try using multiple plugins that require the autoloader, and verify that no errors occur.

#### Proposed changelog entry for your changes:
* N/A

kbrown9 committed:
Autoloader: loop through all map files in enqueue_files()

The latest autoloader needs to loop through all of the classmap and
filemap files created by the new autoloader and populate the globals.
Add a loop through the active plugins to the enqueue_files() function.

zinigor commented:
It's working for me! I have tried to do the same thing I tried while testing the previous PR, and the earlier autoloader picks up the later perfectly fine. Even when I tried to instantiate an autoloaded class before the other plugin has even loaded, it worked! I don't know if this is too much black magic (should we even be allowing that, I mean), but it's definitely better than what we've had before.

kbrown9 committed:
Autoloader: improve enqueue_files

Improve enqueue_files() by:
 - checking the map files with is_readable instead of file_exists
 - checking that the local map variables are arrays before using them
 - move the file_loader() block out of the active plugins loop

kbrown9 committed:
Autoloader: remove plugins_loaded action conditional around file_loader

We no longer need to wait until the plugins_loaded action to call file_loader(),
so remove the conditionals and the plugins_loaded hook.

kbrown9 committed:
Autoloader: Handle an activating plugin

When a plugin is in the process of activating, it won't be in the active
plugins list yet. We need to make sure that its classmap and filemap files
are processed and the globals are updated so that the autoloader can use
them with the newly activated plugin.

 - In enqueue_files(), process the current plugin's map file and the map files
   for all active plugins.
 - In set_up_autoloader(), detect if a new autoloader has been loaded by an
   activating plugin. If it has, reset the global classmap.

kbrown9 committed:
Autoloader: handle an activating plugin

When a plugin is in the process of activating, it's not in the active
plugins list. We need to add the activating plugin's packages to the
autoloader before the plugin loads.

kbrown9 commented on packages/autoloader/src/functions.php:
```diff
@@ -134,38 +134,111 @@ function autoloader( $class_name ) {
+	if ( isset( $_REQUEST['action'] ) && 'activate' === $_REQUEST['action'] ) {
+		$activating_plugin = isset( $_REQUEST['plugin'] ) ? wp_unslash( $_REQUEST['plugin'] ) : null;
+		return $activating_plugin;
+	}
```
This probably needs a nonce verification.

zinigor replied:
You're right, but it turns out that the function that checks the validity of the nonce is not yet loaded at that time. So I have added a "check" that only requires the nonce to be present in the request, but doesn't actually check it. The request will fail later when the plugin is activated.

kbrown9 replied:
This looks good to me!

zinigor committed:
Added a check against bulk activation.

zinigor committed:
Added network activated plugins to the active plugin list.

zinigor committed:
Added a nonce 'check'.
